File: frontend/src/services/booking.ts
import api from "./api";


// ---------------------------------
// Booking Interface
// ---------------------------------

export interface Booking {

    id: string;

    booking_number: string;

    booking_date: string;


    customer_id: string;

    broker_id: string;

    vehicle_id: string;


    loading_site_id: string;

    unloading_site_id: string;


    bill_to?: string;


    weight: number;

    freight_type?: string;



    customer_freight: number;

    customer_advance: number;

    customer_balance: number;

    customer_paid_amount: number;

    tds_amount: number;

    customer_payment_status: "pending" | "partial" | "paid";



    broker_freight: number;

    broker_advance: number;

    broker_balance: number;

    broker_paid_amount: number;

    broker_payment_status: "pending" | "partial" | "paid";



    payment_method?: string;

    bill_submission_status?: string;


    payment_received_date?: string;

    payment_received_amount: number;



    profit: number;


    remarks?: string;


    created_at: string;

    updated_at: string;

}



// ---------------------------------
// Create Booking Payload
// ---------------------------------

export interface CreateBookingPayload {


    booking_date: string;


    customer_id: string;

    broker_id: string;

    vehicle_id: string;



    loading_site_id: string;

    unloading_site_id: string;



    bill_to?: string;



    weight: number;

    freight_type?: string;



    customer_freight: number;

    customer_advance?: number;

    tds_amount?: number;



    broker_freight: number;

    broker_advance?: number;



    payment_method?: string;


    bill_submission_status?: string;



    payment_received_date?: string;


    payment_received_amount?: number;



    remarks?: string;

}



// ---------------------------------
// Update Booking Payload
// ---------------------------------

export interface UpdateBookingPayload {


    booking_date?: string;


    customer_id?: string;

    broker_id?: string;

    vehicle_id?: string;



    loading_site_id?: string;

    unloading_site_id?: string;



    bill_to?: string;



    weight?: number;

    freight_type?: string;



    customer_freight?: number;

    customer_advance?: number;

    tds_amount?: number;



    broker_freight?: number;

    broker_advance?: number;



    payment_method?: string;


    bill_submission_status?: string;



    payment_received_date?: string;


    payment_received_amount?: number;



    remarks?: string;

}





// ---------------------------------
// Get All Bookings
// ---------------------------------

export const getBookings = async () => {


    const response = await api.get<Booking[]>(
        "/bookings/"
    );


    return response.data;

};





// ---------------------------------
// Get Booking By ID
// ---------------------------------

export const getBookingById = async (
    id: string
) => {


    const response = await api.get<Booking>(
        `/bookings/${id}`
    );


    return response.data;

};





// ---------------------------------
// Create Booking
// ---------------------------------

export const createBooking = async (
    data: CreateBookingPayload
) => {


    const response = await api.post<Booking>(

        "/bookings/",

        data

    );


    return response.data;

};





// ---------------------------------
// Update Booking
// ---------------------------------

export const updateBooking = async (

    id: string,

    data: UpdateBookingPayload

) => {


    const response = await api.put<Booking>(

        `/bookings/${id}`,

        data

    );


    return response.data;

};





// ---------------------------------
// Delete Booking
// ---------------------------------

export const deleteBooking = async (

    id: string

) => {


    const response = await api.delete(

        `/bookings/${id}`

    );


    return response.data;

};

File: backend/app/schemas/booking.py
import uuid

from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import (
    BaseModel,
    ConfigDict,
)



# -------------------------
# Create Booking
# -------------------------

class BookingCreate(BaseModel):

    booking_date: date


    customer_id: uuid.UUID

    broker_id: uuid.UUID

    vehicle_id: uuid.UUID



    # Site References

    loading_site_id: uuid.UUID

    unloading_site_id: uuid.UUID



    bill_to: Optional[str] = None



    weight: Decimal

    freight_type: Optional[str] = None



    # Customer Financials

    customer_freight: Decimal

    customer_advance: Decimal = Decimal("0.00")

    tds_amount: Decimal = Decimal("0.00")



    # Broker Financials

    broker_freight: Decimal

    broker_advance: Decimal = Decimal("0.00")



    # Payment

    payment_method: Optional[str] = None

    bill_submission_status: Optional[str] = None


    payment_received_date: Optional[date] = None


    payment_received_amount: Decimal = Decimal("0.00")



    remarks: Optional[str] = None





# -------------------------
# Update Booking
# -------------------------

class BookingUpdate(BaseModel):

    booking_date: Optional[date] = None



    customer_id: Optional[uuid.UUID] = None

    broker_id: Optional[uuid.UUID] = None

    vehicle_id: Optional[uuid.UUID] = None



    # Site References

    loading_site_id: Optional[uuid.UUID] = None

    unloading_site_id: Optional[uuid.UUID] = None



    bill_to: Optional[str] = None



    weight: Optional[Decimal] = None

    freight_type: Optional[str] = None



    customer_freight: Optional[Decimal] = None

    customer_advance: Optional[Decimal] = None

    tds_amount: Optional[Decimal] = None



    broker_freight: Optional[Decimal] = None

    broker_advance: Optional[Decimal] = None



    payment_method: Optional[str] = None

    bill_submission_status: Optional[str] = None


    payment_received_date: Optional[date] = None


    payment_received_amount: Optional[Decimal] = None



    remarks: Optional[str] = None





# -------------------------
# Response
# -------------------------

class BookingResponse(BaseModel):

    id: uuid.UUID



    booking_number: str

    booking_date: date



    customer_id: uuid.UUID

    broker_id: uuid.UUID

    vehicle_id: uuid.UUID



    # Site References

    loading_site_id: uuid.UUID

    unloading_site_id: uuid.UUID



    bill_to: Optional[str]



    weight: Decimal

    freight_type: Optional[str]



    customer_freight: Decimal

    customer_advance: Decimal

    customer_balance: Decimal

    customer_paid_amount: Decimal

    tds_amount: Decimal

    customer_payment_status: str



    broker_freight: Decimal

    broker_advance: Decimal

    broker_balance: Decimal

    broker_paid_amount: Decimal

    broker_payment_status: str



    payment_method: Optional[str]

    bill_submission_status: Optional[str]



    payment_received_date: Optional[date]

    payment_received_amount: Decimal



    profit: Decimal



    remarks: Optional[str]



    created_at: datetime

    updated_at: datetime



    model_config = ConfigDict(
        from_attributes=True
    )
